app/controllers/application_controller.rb

```diff
@@ -1,10 +1,149 @@
-require './config/environment'
-
 class ApplicationController < Sinatra::Base
-
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if session[:id]
+      redirect '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if empty_fields?
+      redirect '/signup'
+    else
+      @user = User.create(params)
+      session[:id] = @user.id
+      redirect '/tweets'
+    end
+  end
+
+  get '/tweets' do
+    if logged_in?
+      @user = User.find_by(id: session[:id])
+      erb :'tweets/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/users/:slug' do
+    # binding.pry
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  get '/tweets/new' do
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do
+    if !params[:content].empty?
+      @user = current_user
+      @tweet = @user.tweets.build(content: params[:content])
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/edit_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+    if logged_in?
+      if !params[:content].empty?
+        @tweet.content = params[:content]
+        @tweet.save
+        redirect "/tweets/#{@tweet.id}"
+      else
+      # binding.pry
+        redirect "/tweets/#{@tweet.id}/edit"
+      end
+    end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+    if user && user.authenticate(params[:password])
+      session[:id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/failure"
+    end
   end
 
-end
\ No newline at end of file
+  # added extra logic for logged_in?
+  delete '/tweets/:id/delete' do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      if @tweet.user_id == current_user.id
+        @tweet.delete
+      end
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/logout' do
+    session.clear
+    redirect '/login'
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:id]
+    end
+
+    def current_user
+      User.find(session[:id])
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)).